Block Hooks API: Move ignoredHookedBlocks metadata injection logic

tjcafferkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/60759

**Testing instructions:**

1. Add the below code to your site.
2. Load the navigation block and check it renders in the editor/frontend
3. In the editor, update the navigation by adding another page and saving
4. Check the `wp_postmeta` table to see if `core/loginout` has been added to a row with the `meta_key` => `_wp_ignored_hooked_blocks`

```php
function register_logout_block_as_navigation_last_child( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( $anchor_block === 'core/navigation' && $position === 'last_child' ) {
		$hooked_blocks[] = 'core/loginout';
	}

	return $hooked_blocks;
}

add_filter( 'hooked_block_types', 'register_logout_block_as_navigation_last_child', 10, 4 );
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

tjcafferkey commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+	 * Skip meta generation when consumers intentionally update specific Navigation fields and omit the content update.
+	 * Or if the incoming post is not a Navigation post type.
+	 */
+	if ( ! isset( $post->post_content ) || ( isset( $post->post_type ) && 'wp_navigation' !== $post->post_type ) ) {
```
Probably makes sense to combine the `post_content` and the `ID` check above it into a single statement.

ockham replied:
Or even put it separately? I don't hate how [we'd organized it in GB](https://github.com/WordPress/gutenberg/blob/6108134aae75d1bd4826256490c609cb29044cd8/packages/block-library/src/navigation/index.php#L1572-L1578), as each case is fairly different in nature, and having separate comments kinda makes sense IMO 🤔

tjcafferkey replied:
https://github.com/WordPress/wordpress-develop/pull/6604/commits/7878d8a84f48ce9c9eff873611575169d7efdeb8

tjcafferkey commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+ * @param stdClass $post Post object.
+ * @return stdClass The updated post object.
+ */
+function update_ignored_hooked_blocks_postmeta( $post ) {
```
I've tried to keep this function generic enough for us to adapt it to support other blocks/post types, but at the same time being somewhat specific to `wp_navigation`

ockham replied:
Love the name BTW!

tjcafferkey commented on src/wp-includes/default-filters.php:
```diff
@@ -757,4 +757,7 @@
 add_filter( 'rest_pre_insert_wp_template_part', 'inject_ignored_hooked_blocks_metadata_attributes' );
 
+// Update ignoredHookedBlocks postmeta for wp_navigation post type.
+add_filter( 'rest_pre_insert_wp_navigation', 'update_ignored_hooked_blocks_postmeta' );
```
I will need to add a `has_filter()` check for this in Gutenberg

ockham replied:
I've heard that the Gutenberg version that will be used to sync code into WP 6.6 will be an RC version that's going to be published next Friday (31st). I think that should be enough time to get the GB PR ready and merged, right? It's basically just that `has_filter` check IIUC 😊

tjcafferkey replied:
https://github.com/WordPress/gutenberg/pull/61903

tjcafferkey commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+ *                                                       or pattern that the blocks belong to.
+ * @return string The serialized markup with the ignored hooked blocks metadata applied.
+ */
+function apply_hooked_blocks_to_content( $content, $context, $callback = 'insert_hooked_blocks' ) {
```
Open to renaming this function.

ockham replied:
How about `insert_hooked_blocks_into_content`? Or `inject_`?

We already have `insert_hooked_blocks` (which is the callback used by `make_*_block_visitor`). So maybe `insert_` for consistency's sake 🤔

_Edit:_ Oops, it's literally also an argument to _this_ function.

tjcafferkey replied:
https://github.com/WordPress/wordpress-develop/pull/6604/commits/1cd5fbee261bda2593b4223c08d1c91a1b51f9b1

tjcafferkey commented:
cc @ockham

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,115 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+		}
+	}
+
+	$mocked_parent_block = array(
+		'blockName'    => 'core/navigation',
+		'attrs'        => $attributes,
+		'innerBlocks'  => $blocks,
+		'innerContent' => array_fill( 0, count( $blocks ), null ),
+	);
+
+	$serialized_block = apply_hooked_blocks_to_content( traverse_and_serialize_block( $mocked_parent_block ), get_post( $post->ID ), 'set_ignored_hooked_blocks_metadata' );
+	$root_block       = parse_blocks( $serialized_block )[0];
```
We'll need to optimize this a bit 😬 Along with [this `parse_blocks` call](https://github.com/WordPress/wordpress-develop/pull/6604/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R1074), we're:

1. Parsing `post_content` (to attach the resulting blocks to the mocked parent block),
2. traversing and reserializing it,
3. calling `apply_hooked_blocks_to_content`, which
    1. parses the serialized markup, and
    2. traverses and reserializes it
4. parsing the resulting markup (to get the root block).

Since parsing and traversal/serialization are expensive operations, let's try to cancel out as many of them as possible. Roughly speaking, I think we can construct the mock parent block without needing to parse `post_content` at that stage, but instead do it on a markup base that we then feed to `apply_hooked_blocks_to_content`.

I think the `get_comment_delimited_block_content()` function should help with that. Something like

```php
$markup = get_comment_delimited_block_content(
	'core/navigation',
	$attributes,
	$post->post_content
);
$serialized_block = apply_hooked_blocks_to_content( $markup, get_post( $post->ID ), 'set_ignored_hooked_blocks_metadata' );
```

(untested)

tjcafferkey replied:
Nice, this does work as expected https://github.com/WordPress/wordpress-develop/pull/6604/commits/53033c4d015b869df9865b3bfc602782646eefd1

I'd love to optimize this further if I can find opportunity to do so. I'm going to focus on some tests to support it before doing so though.

ockham commented:
This is shaping up nicely! 😄

ockham commented:
We have some test coverage in https://github.com/WordPress/gutenberg/blob/6108134aae75d1bd4826256490c609cb29044cd8/phpunit/blocks/block-navigation-block-hooks-test.php that we can probably copy over (with some small tweaks).

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,120 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+ *                                                       or pattern that the blocks belong to.
+ * @return string The serialized markup with the ignored hooked blocks metadata applied.
+ */
+function insert_hooked_blocks_into_content( $content, $context, $callback = 'insert_hooked_blocks' ) {
```
I need to retract my earlier naming suggesting 😭 I realized that the function's behavior is dependent on the `$callback`, so it doesn't always insert hooked blocks.

I guess "applying" something "recursively" to "content", or to a "block tree", are better keywords here. We can revert back to `apply_hooked_blocks_to_content`. Or maybe `apply_block_hooks_to_content`? (As "Block Hooks" is the algorithm we're applying; whereas hooked blocks we aren't always adding/applying) 🤔

tjcafferkey replied:
Yeah, good catch as that had totally bypassed me too. Updated in https://github.com/WordPress/wordpress-develop/pull/6604/commits/08e459bb237c8540b6e3c337c9ddb4c92587bdb6

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,123 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+ * @access private
+ *
+ * @param string $serialized_block The serialized markup of a block and its inner blocks.
+ * @return string
```
```suggestion
 * @return string The serialized markup of the inner blocks.
```

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,123 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
 }
 
+/**
+ * Runs the hooked blocks algorithm on the given content
```
Nitpick

```suggestion
 * Runs the hooked blocks algorithm on the given content.
```

ockham replied:
809ec8259f12a4efef4a36e6159f5f4b0df8e7bf

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,123 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+
+	$attributes = array();
+
+	if ( isset( $post->ID ) ) {
```
We can remove this check, since [we bail early if `$post->ID` is empty](https://github.com/WordPress/wordpress-develop/pull/6604/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R1065-R1067).

ockham replied:
657dfb51a54f0059b38a94dc943df6f6a1696d73

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,123 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+		$post->post_content
+	);
+
+	$serialized_block = apply_block_hooks_to_content( $markup, get_post( $post->ID ), 'set_ignored_hooked_blocks_metadata' );
```
The `$context` here will reflect what's currently in the DB, but we should be passing the new content that's coming in over the wire. This is a similar problem to [the one we had with templates](https://core.trac.wordpress.org/ticket/60754). I don't think we should do it as part of this PR but separately; I'll just leave some notes here for now.

The fix more straight-forward here, as we don't have to map a `$changes` `stdClass` instance to a `WP_Block_Template` type, but only to a `WP_Post` object. We'll still need to apply the `$changes` on top of the existing `wp_navigation` post object.

Something like the following (inspired by [this](https://github.com/WordPress/wordpress-develop/pull/6278/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7R1510-R1522) -- minus the revision handling, which we don't need):

```suggestion
	// Find the existing post object.
	$existing_post = get_post( $post->ID );

	// Apply the changes to the existing post object.
	$post = (object) array_merge( (array) $existing_post, (array) $post );

	$serialized_block = apply_block_hooks_to_content( $markup, $post, 'set_ignored_hooked_blocks_metadata' );
```

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+	 * @covers ::update_ignored_hooked_blocks_postmeta
+	 */
+	public function test_block_core_navigation_update_ignore_hooked_blocks_meta_preserves_entities() {
+		if ( ! function_exists( 'set_ignored_hooked_blocks_metadata' ) ) {
+			$this->markTestSkipped( 'Test skipped on WordPress versions that do not included required Block Hooks functionality.' );
+		}
```
We don't have that excuse here -- we're in Core, where we know that the function does exist 😄 

```suggestion
		if ( ! function_exists( 'set_ignored_hooked_blocks_metadata' ) ) {
			$this->markTestSkipped( 'Test skipped on WordPress versions that do not included required Block Hooks functionality.' );
		}
```

tjcafferkey replied:
>We don't have that excuse here -- we're in Core, where we know that the function does exist 😄

Thanks, I wasn't sure whether these also run against previous WP versions for back-compat reasons.

ockham replied:
572ed7102de9ee0bb9c0af98975106219e1b24ad

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+	/**
+	 * @covers ::update_ignored_hooked_blocks_postmeta
+	 */
+	public function test_block_core_navigation_update_ignore_hooked_blocks_meta_preserves_entities() {
```
The method names still reflect the "old" filter name, we should update them.

```suggestion
	public function test_update_ignored_hooked_blocks_postmeta_preserves_entities() {
```

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -1002,6 +1002,123 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
+	/**
+	 * Skip meta generation when the post content is not a navigation block.
+	 */
+	if ( ! isset( $post->post_type ) || 'wp_navigation' !== $post->post_type ) {
```
(I deleted a comment that I wrote earlier where I was worried that `post_type` might not always be set. Turns out that [it always is set](https://github.com/WordPress/wordpress-develop/blob/6f6bd30d37360670cc1daa96d08dcbac2f038c07/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1235-L1242). All good.)

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+		);
+	}
+
+		/**
```
```suggestion
	/**
```

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+ * @group blocks
+ */
+class Tests_Blocks_UpdateIgnoredHookedBlocksPostMeta extends WP_UnitTestCase {
+
+	/**
+	 * Original markup.
+	 *
+	 * @var string
+	 */
+	protected static $original_markup;
```
We're currently not using this. We should either use it (and remove the local vars from each test; though one of them seems to slightly differ from the others), or remove it.

ockham replied:
34526ff2505e1a757f34228215ef3e5f395956c7

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+	 * Setup method.
+	 */
+	public static function wpSetUpBeforeClass() {
+		//self::$original_markup = '<!-- wp:navigation-link {"label":"News & About","type":"page","id":2,"url":"http://localhost:8888/?page_id=2","kind":"post-type"} /-->';
```
See https://github.com/WordPress/wordpress-develop/pull/6604/files#r1612007218 :)

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+ *
+ * @since 6.6.0
+ *
+ * @group blocks
```
I saw the other day that for a test class where all methods cover the same function, we can do

```suggestion
 * @group blocks
 * @covers ::update_ignored_hooked_blocks_postmeta
```

and remove it from the individual methods 😄

ockham replied:
Fixed by @tjcafferkey in 79b9bedc22c424744a239af040181ea911936d6c.

ockham commented on tests/phpunit/tests/blocks/updateIgnoredHookedBlocksPostMeta.php:
```diff
@@ -0,0 +1,225 @@
+		parent::tear_down();
+	}
+
+	/**
+	 * @covers ::update_ignored_hooked_blocks_postmeta
+	 */
```
I think we should add ticket numbers to the methods (but can remove the `@covers` if we do [this](https://github.com/WordPress/wordpress-develop/pull/6604/files#r1612010086) instead).

```suggestion
	/**
	 * @ticket 60759
	 */
```

ockham replied:
https://github.com/WordPress/wordpress-develop/commit/79b9bedc22c424744a239af040181ea911936d6c

ockham reviewed (commented):
Thank you very much, this is looking great! 🚀 Especially appreciate the test coverage, good thinking to add test cases for missing post ID and type 👍 

Apologies for the number of notes I left, most of it is minor and/or coding standards compliance stuff 😬

tjcafferkey commented:
@ockham I've addressed all your comments aside from the one we're going to address in a separate PR. I've also had to fix some merge conflicts with https://github.com/WordPress/wordpress-develop/pull/6358

ockham commented:
> @ockham I've addressed all your comments aside from the one we're going to address in a separate PR. I've also had to fix some merge conflicts with #6358

Awesome, thank you very much!

We'll need to land the [GB counterpart PR](https://github.com/WordPress/gutenberg/pull/61903) first; I've approved it, you wanna do the honors?

After that, we'll still need to wait for the code to be sync'd over to Core, so that the [built Navigation block code in Core](https://github.com/WordPress/wordpress-develop/blob/1025e34d4f7eb93ccfacfbd98f750f02688bfdff/src/wp-includes/blocks/navigation.php#L1619-L1625) is updated and the "old" filter isn't still being added there if the new one is present. As always, the package sync will happen before Feature Freeze (i.e. next Tue, June 4), but probably only after the GB RC is released (this Friday). This means we'll likely land this PR on Monday 🙂

ockham commented:
> We'll need to land the [GB counterpart PR](https://github.com/WordPress/gutenberg/pull/61903) first; I've approved it, you wanna do the honors?

I remembered you won't be able to do so today, so I want ahead and merged it 😬 Hope that's okay!

ockham committed:
Block Hooks API: Move ignoredHookedBlocks metadata injection logic

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/58291.
